### PR TITLE
HDDS-3510. Upgrade Ratis to 0.6.0-2816ea6-SNAPSHOT

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Upgrade Ratis to `0.6.0-2816ea6-SNAPSHOT` to get the fix for [RATIS-840](https://issues.apache.org/jira/browse/RATIS-840).

https://issues.apache.org/jira/browse/HDDS-3510

## How was this patch tested?

Two successful CI runs:

https://github.com/adoroszlai/hadoop-ozone/runs/632673436
https://github.com/adoroszlai/hadoop-ozone/runs/635512527

Note that `TestReadRetries.testPutKeyAndGetKeyThreeNodes` failed in two other CI runs:

https://github.com/adoroszlai/hadoop-ozone/runs/629899380
https://github.com/adoroszlai/hadoop-ozone/runs/632680463

but it also failed on `master` already:

https://github.com/apache/hadoop-ozone/runs/626422906

created [HDDS-3516](https://issues.apache.org/jira/browse/HDDS-3516) for `TestReadRetries`.

### CODE REVIEW

**Comment by elek** [comment]
Why did `TestKeyInputStream.testSeek` fail in build https://github.com/apache/hadoop-ozone/pull/890/commits/e1ddffddd0e0784f1fec802d0348b67bd10d408c?

Couldn't see the same error in master. Can it be related to the version bump?

**Comment by adoroszlai** [comment]
> Why did `TestKeyInputStream.testSeek` fail in build [e1ddffd](https://github.com/apache/hadoop-ozone/commit/e1ddffddd0e0784f1fec802d0348b67bd10d408c)?
> 
> Couldn't see the same error in master. Can it be related to the version bump?

It could be.  I've [seen](https://github.com/adoroszlai/hadoop-ozone/runs/632572212) this in [another branch](https://github.com/adoroszlai/hadoop-ozone/commits/CI_with_RATIS-840) where I tried to enable more integration tests with this Ratis version.  Haven't happened since adding a check (to throw exception on `0 -> -1` op. count transition.

`TestCloseContainerByPipeline.testQuasiCloseTransitionViaRatis` seems to be [timing out](https://github.com/adoroszlai/hadoop-ozone/runs/635919602) more frequently.

**Comment by elek** [comment]
@lokeshj1703 Do you have any idea about the `TestKeyInputStream.testSeek` failure? We haven't seen it on master. Is it related to the version bump?

**Comment by lokeshj1703** [comment]
@elek It might be related to HDDS-3223 which added copyLarge function in KeyInputStream and added a corresponding test in TestKeyInputStream.

**Comment by elek** [comment]
> @elek It might be related to HDDS-3223 which added copyLarge function in KeyInputStream and added a corresponding test in TestKeyInputStream.

How is it related? Is this problem introduced or fixed by HDDS-3223?

* If introduced, we need a revert. But I can't see to high possibility as we didn't see this error since the merge of HDDS-3223 

* If fixed, it supposed to be disappeared after the patch (HDDS-3223 was part of the failing test).

In both cases: HDDS-3223 merged 7 days ago. Any intermittent failure related to that patch should be visible on master either after or before the merge.

**Comment by bshashikant** [comment]
> TL;DR;
> 
> One of the CI builds failed with a new type of test failure which was not seen before on master.
> 
> To make sure that no new intermittent failure is introduced we need more testing and analyses of the failures (IMHO) with the help of Ratis experts....

I will look into it. Can you point me to the logs for failed case?

**Comment by lokeshj1703** [comment]
> How is it related? Is this problem introduced or fixed by [HDDS-3223](https://issues.apache.org/jira/browse/HDDS-3223)?
> 

The failure occurs due to metric count mismatch. The problem seems to be happening due to wrong metric being compared. Created jira https://issues.apache.org/jira/browse/HDDS-3549 for tracking the issue. We can merge this jira and then see if the test failure occurs again.

**Comment by elek** [comment]
New failure:

```
[INFO] Running org.apache.hadoop.ozone.dn.ratis.TestDnRatisLogParser
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 19.261 s <<< FAILURE! - in org.apache.hadoop.ozone.dn.ratis.TestDnRatisLogParser
[ERROR] testRatisLogParsing(org.apache.hadoop.ozone.dn.ratis.TestDnRatisLogParser)  Time elapsed: 19.165 s  <<< FAILURE!
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at org.apache.hadoop.ozone.dn.ratis.TestDnRatisLogParser.testRatisLogParsing(TestDnRatisLogParser.java:84)
```

**Comment by lokeshj1703** [comment]
> BTW, I didn't fully understand how HDDS-3549 can be related.

So TestKeyInputStream#testSeek was comparing readChunkCount with metric for pendingReadChunks. So that is a problem. It can be a possible reason for metric count to not match. 
> Is it an intermittent test? Why didn't we see it earlier?

The reason I mentioned HDDS-3223 is because it adds a new test which can lead to change in metric values since it would also do read chunks. I don't have an exact explanation of how this could happen though. But for concluding we need to fix the test in TestKeyInputStream and then see if it still fails.

**Comment by adoroszlai** [comment]
FYI first `testSeek` failure on `master`: https://github.com/apache/hadoop-ozone/runs/654811334

**Comment by mukul1987** [comment]
This is getting duplicated by #911. Closing this for now.
